crates/alloy-provider/src/lib.rs

```diff
@@ -0,0 +1,1460 @@
+    }
+}
+
+impl<P, Node: NodeTypes, N> AlloyRethProvider<P, Node, N> {
+    /// Creates a new `AlloyRethProvider` with default configuration
+    pub fn new(provider: P) -> Self
+    where
+        Node::ChainSpec: Default,
+    {
+        Self::new_with_config(provider, AlloyRethProviderConfig::default())
+    }
+
+    /// Creates a new `AlloyRethProvider` with custom configuration
+    pub fn new_with_config(provider: P, config: AlloyRethProviderConfig) -> Self
+    where
+        Node::ChainSpec: Default,
+    {
+        let (canon_state_notification, _) = broadcast::channel(1);
+        Self {
+            provider,
+            node_types: std::marker::PhantomData,
+            network: std::marker::PhantomData,
+            canon_state_notification,
+            config,
+            chain_spec: Arc::new(Node::ChainSpec::default()),
+        }
+    }
+
+    /// Helper function to execute async operations in a blocking context
+    fn block_on_async<F, T>(&self, fut: F) -> T
+    where
+        F: Future<Output = T>,
+    {
+        tokio::task::block_in_place(move || Handle::current().block_on(fut))
+    }
+}
+
+impl<P, Node, N> AlloyRethProvider<P, Node, N>
+where
+    P: Provider<N> + Clone + 'static,
+    N: Network,
+    Node: NodeTypes,
+{
+    /// Helper function to create a state provider for a given block ID
+    fn create_state_provider(&self, block_id: BlockId) -> AlloyRethStateProvider<P, Node, N> {
+        AlloyRethStateProvider::with_chain_spec(
+            self.provider.clone(),
+            block_id,
+            self.chain_spec.clone(),
+        )
+    }
+
+    /// Helper function to get state provider by block number
+    fn state_by_block_number(
+        &self,
+        block_number: BlockNumber,
+    ) -> Result<StateProviderBox, ProviderError> {
+        Ok(Box::new(self.create_state_provider(BlockId::number(block_number))))
+    }
+}
+
+// Implementation note: While the types are generic over Network N, the trait implementations
+// are specialized for AnyNetwork because they need to access block header fields.
+// This allows the types to be instantiated with any network while the actual functionality
+// requires AnyNetwork. Future improvements could add trait bounds for networks with
+// compatible block structures.
+impl<P, Node, N> BlockHashReader for AlloyRethProvider<P, Node, N>
+where
+    P: Provider<N> + Clone + 'static,
+    N: Network,
+    Node: NodeTypes,
+{
+    fn block_hash(&self, number: BlockNumber) -> Result<Option<B256>, ProviderError> {
+        let block = self.block_on_async(async {
+            self.provider.get_block_by_number(number.into()).await.map_err(ProviderError::other)
+        })?;
+        Ok(block.map(|b| b.header().hash()))
+    }
+
+    fn canonical_hashes_range(
+        &self,
+        _start: BlockNumber,
+        _end: BlockNumber,
+    ) -> Result<Vec<B256>, ProviderError> {
+        // Would need to make multiple RPC calls
+        Ok(Vec::new())
```

Comment:
maybe better express clearly that this is not supported?
```suggestion
        Err(ProviderError::UnsupportedProvider)
```